lib/normalizeSynonyms.coffee

```diff
@@ -0,0 +1,26 @@
+# Simplistic approach to providing synonym-matching, e.g. in typeahead queries.
+# 
+# Usage: 
+#
+# mySynonyms = [ ['saint', 'st'], ['harbor', 'harbour'] ]
+# normalizeSynonyms(mySynonyms, 'st ives')
+#  => returns 'saint ives'
+# normalizeSynonyms(mySynonyms, 'bal harbour')
+#  => returns 'bal harbor'
+
+normalizeSynonymList = (synonymList, inputString) -> 
+  firstSynonym = synonymList[0]
+  synonymList.reduce (str, synonym) -> 
+    regex = new RegExp '\\b' + synonym + '\\b', 'i'
+    str.replace(regex, firstSynonym)
```

Comment:
For each list of synonyms, we derive the final, "synonymized" string by doing a `reduce` operation on the list of synonyms, using the input string as the initial value to the accumulator, and replacing each synonym with a common term (the first-provided synonym, in this case)